### PR TITLE
Lms metadata 3

### PR DESCRIPTION
**Pull request by SteveMicroNova**
### What does this change intend to accomplish?
Finally fix and implement the LMS metadata collection script I originally made just over a year ago
 
### Checklist

* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?
* [x] If applicable, have you updated the documentation/manual?
* [x] If applicable, have you updated the CHANGELOG?
* [x] Does your submission pass linting & tests? You can test on localhost using `./scripts/test`
* [ ] Have you written new tests for your core features/changes, as applicable?

### CODE REVIEW

**Comment by SteveMicroNova** [comment]
Currently it's incredibly efficient at finding the player when there isn't a specified server; python's socket module seems to be complaining whenever I try to resolve the hostname but also that's because the LMS server is on my windows machine and the metadata script runs in WSL and the hostname resolves to the WSL instance instead of windows; needs more testing outside of my super specific setup to verify that it's good with or without hostname being provided

Multiple servers/players on the same network seem to be a trouble right now as well, but that's a less common usecase and not the highest order concern while I figure out the single stream single server setup.

**Comment by SteveMicroNova** [comment]
This feels ready for testing

To test, spin up multiple LMS streams with various settings on one AmpliPro after deploying this branch to it and play things to each instance and witness the metadata. If you need a unit to test on, lamplipi (the one on lincolns desk) currently has it deployed

I don't know how to / if I have to implement that into our pytests thing as the last checkbox of the PR template suggest

File: streams/lms_metadata.py
Comment (rtertiaer) [inline streams/lms_metadata.py:90]: i'd love to see this more effectively case-statemented out; I can see a world where we run both x86_64, arm, and arm64.

is there a fallback for when these executables cannot be run? it doesn't appear we actually handle the exception potentially generated from `check=true`; that feels important here.

**Review by rtertiaer — CHANGES_REQUESTED** [review]
how did you generate the binaries for this?

**Comment by SteveMicroNova** [comment]
> how did you generate the binaries for this?

They were compiled a year ago so I don't fully remember

I'm pretty sure I compiled the find_lms_server.c once on my local (x64) and once on an amplipi (arm)

**Comment by SteveMicroNova** [comment]
I don't know how to compile C programs for other architectures so I've simply made it an error case if there's an unsupported architecture that tries to run the program
Ideally if we ever use a Pi with a different type of CPU, we'll just compile all of our binaries on it as well so we'll have the right ones available before sending it out into the world

File: streams/lms_metadata.py
Comment (SteveMicroNova) [reply streams/lms_metadata.py]: Now handled explicitly with `except subprocess.CalledProcessError` in recent commit

File: streams/lms_metadata.py
Comment (rtertiaer) [inline streams/lms_metadata.py:113]: what is the "less fast" way of connecting to the metadata server? is it an acceptable fallback for the unsupported CPU arch case? ie, is it pure Python?

Comment (SteveMicroNova) [reply streams/lms_metadata.py:113]: The less fast way is via manually scraping the whole network with python nmap; one that was prohibitively slow and also annoyed the hell out of every other device on the network by going door to door and asking "Are you an LMS server?"

Technically possible, doesn't seem worthwhile or sensible

IMO it'd be better to implement the ability to write down the IP as this script already supports that and is simply awaiting the LMS stream to add it in as an option during stream creation

Comment (rtertiaer) [reply streams/lms_metadata.py:113]: > IMO it'd be better to implement the ability to [input] the IP

after reading through the rest of this, I tend to agree. if you:
a) haven't provided a hostname or IP
b) the LMS server cannot be autodiscovered, with the binary provided
then you've already come to the end of the road. making (a) accept more inputs seems like a reasonable path here

**Comment by SteveMicroNova** [comment]
I don't know why MyPy linting is complaining, the line it's currently complaining was just reverted to the way it's almost always been until like three commits ago

File: amplipi/streams.py
Comment (rtertiaer) [inline amplipi/streams.py:1345]: it looks like the thing mypy is complaining about is the `Optional` bit; i think it may be because self.name and self.server are not guaranteed to be set here, which could indeed trigger a bug

**Comment by SteveMicroNova** [comment]
I'd like to add port support during this PR, but I don't know how to do so
That requires changing the stream; I've found out how to add the spot in the form on the frontend but I don't know how to make that data actually be saved when it's input so far

File: amplipi/streams.py
Comment (SteveMicroNova) [inline amplipi/streams.py:1348]: This makes me feel like port should be used around here
I don't currently know where or how
I mean, it has to be appended like `lms_args += ['-s', server:port]`, but the use of "assumed" makes me feel that 9000 is input somewhere somehow automatically and I'll have to replace it first

**Comment by SteveMicroNova** [comment]
This actually needs more work before merging
We have proper folders to store different metadata in, Lincoln told me about it verbally and I haven't implemented that yet because I got focused on making the branch actually work and forgot to make it work good
I'm currently on #667, I'll loop back to this in a bit

File: streams/lms_metadata.py
Comment (SteveMicroNova) [inline streams/lms_metadata.py:217]: Left local test changes in on accident, these will be cleaned out in next commit

File: streams/lms_metadata.py
Comment (SteveMicroNova) [inline streams/lms_metadata.py:132]: image

For some reason if you don't populate self.server during __init__, this has a chance of never getting anything for that value and so polls for `http://:900/jsonrpc.js` which is a fail state

self.server should always be populated by this point though, if it wasn't solved by `resolve_host` then it should be solved by `find_lms_server`; if somehow neither of those happen it should error out before ever reaching this point in the code

**Comment by SteveMicroNova** [comment]
The metadata script is incredibly clean and efficient, if noisy at the moment
I'm moving on to more fruitful endeavors for now as Ryan and I have identified an issue I've been having as inherent to our alsaloop implementation or the version of LMS we're running, either way something that is out of my power and knowledge and may need work done outside of this branch to solve

**Comment by SteveMicroNova** [comment]
Issue #679 that was blocking this PR has been solved in #681, this branch now works and is safe to review and eventually merge :+1:

File: amplipi/models.py
Comment (rtertiaer) [inline amplipi/models.py:86]: Is it reasonable to set a default here?

File: amplipi/models.py
Comment (rtertiaer) [inline amplipi/models.py:110]:
```suggestion

```

File: amplipi/streams.py
Comment (rtertiaer) [inline amplipi/streams.py:1368]: can you communicate/poll/wait here to prevent zombies?

File: scripts/configure.py
Comment (rtertiaer) [inline scripts/configure.py:161]: `squeezelite` is already installed on these boxes; can you clarify what versions we are going from and to with this?

**Review by rtertiaer — CHANGES_REQUESTED** [review]
Couple of notes, no huge lifts I don't think. Can you also pre-squash this before the next review?

File: CHANGELOG.md
Comment (SteveMicroNova) [reply CHANGELOG.md:9]: We don't have an example of a subheader / subsection, it does seem like we prefer to make new headers rather than having subheaders within what would otherwise be the relevant section
Change made in upcoming commit

File: scripts/configure.py
Comment (SteveMicroNova) [reply scripts/configure.py:161]: I continue to not remember why this line was added
I'm going to simply remove it as the specific LMS / squeezelite install is managed elsewhere, testing just now and removing this line changes nothing

**Comment by SteveMicroNova** [comment]
Turning down the logging level of lms_metadata.py before anything else, logs are absolutely destroyed by that especially if you're in LMS mode with all 4 streams going

**Review by rtertiaer — CHANGES_REQUESTED** [review]
can you squash this?

File: amplipi/streams.py
Comment (rtertiaer) [inline amplipi/streams.py:49]:
```suggestion
sh = logging.StreamHandler(sys.stdout)
```

**Review by rtertiaer — CHANGES_REQUESTED** [review]

**Review by rtertiaer — APPROVED** [review]

**Pull request merged**
